controllers/nstemplateset/nstemplateset_controller.go

```diff
@@ -232,7 +238,75 @@ func listByOwnerLabel(username string) runtimeclient.ListOption {
 	return runtimeclient.MatchingLabels(labels)
 }
 
-func isUpToDateAndProvisioned(obj metav1.Object, tierTemplate *tierTemplate) bool {
-	return obj.GetLabels()[toolchainv1alpha1.TemplateRefLabelKey] != "" &&
+// isUpToDateAndProvisioned checks if the obj has the correct Template Reference Label.
+// If so, it processes the tier template to get the expected roles and rolebindings and then checks if they are actually present in the namespace.
+func isUpToDateAndProvisioned(obj metav1.Object, tierTemplate *tierTemplate, r *namespacesManager) (bool, error) {
+	isLabelCorrect := obj.GetLabels()[toolchainv1alpha1.TemplateRefLabelKey] != "" &&
 		obj.GetLabels()[toolchainv1alpha1.TemplateRefLabelKey] == tierTemplate.templateRef
+
+	if isLabelCorrect {
+
+		newObjs, err := tierTemplate.process(r.Scheme, obj.GetName(), template.RetainAllButNamespaces)
+		if err != nil {
+			return false, err
+		}
+		processedRoles := []runtimeclient.Object{}
+		processedRoleBindings := []runtimeclient.Object{}
+		roleList := rbac.RoleList{}
+		rolebindingList := rbac.RoleBindingList{}
+		err = r.AllNamespacesClient.List(context.TODO(), &roleList, runtimeclient.InNamespace(obj.GetName()))
+		if err != nil {
+			return false, err
+		}
+
+		err = r.AllNamespacesClient.List(context.TODO(), &rolebindingList, runtimeclient.InNamespace(obj.GetName()))
+		if err != nil {
+			return false, err
+		}
+
+		for _, objt := range newObjs {
+			if objt.GetObjectKind().GroupVersionKind().Kind == "Role" {
+				processedRoles = append(processedRoles, objt)
+			}
+			if objt.GetObjectKind().GroupVersionKind().Kind == "RoleBinding" {
+				processedRoleBindings = append(processedRoleBindings, objt)
+			}
+		}
+		//Check the names of the roles and roleBindings as well
+		for _, role := range processedRoles {
+			if !containsRole(roleList.Items, role) {
+				return false, nil
+			}
+		}
+
+		for _, rolebinding := range processedRoleBindings {
+			if !containsRoleBindings(rolebindingList.Items, rolebinding) {
+				return false, nil
+			}
+		}
+
+		return true, nil
+	}
+
+	return isLabelCorrect, nil
```

Comment:
oh, maybe you could return `false` if `isLabelCorrect` is `false`, before even evaluating all the role and rolebindings?

Reply:
That's what it is doing, it only processes role and rolebinding if isLabelCorrect is true.